### PR TITLE
4555 Create server modules for retrieving client application data

### PR DESCRIPTION
**Pull request by Fbasham**
### Description
Adds domain files (DTO, entity, mapper, repository, service) for retrieving client application data. 

### Related Azure Boards Work Items
[AB#4555](https://dev.azure.com/DTS-STN/1fc40a8f-28cf-47bc-b6e4-1c234bd06177/_workitems/edit/4555)

### Checklist

- [x] I have tested the changes locally
- [ ] I have updated the documentation if necessary
- [ ] I have added/updated tests that prove my fix is effective or that my feature works
- [x] I have checked that my code follows the project's coding style by running `npm run format:check`
- [x] I have checked that my code contains no linting errors by running `npm run lint`
- [x] I have checked that my code contains no type errors by running `npm run typecheck`
- [x] I have checked that all unit tests pass by running `npm run test:unit -- run`
- [x] I have checked that all e2e tests pass by running `npm run test:e2e`

### CODE REVIEW

File: frontend/app/.server/container.init.ts [hunk 1]
```diff
@@ -2,6 +2,8 @@ import { error } from 'console';
 import { Container } from 'inversify';
 import { makeLoggerMiddleware, textSerializer } from 'inversify-logger-middleware';
 
+import type { ClientApplicationDtoMapper } from './domain/mappers/client-application.dto.mapper';
+import { ClientApplicationDtoMapperImpl } from './domain/mappers/client-application.dto.mapper';
```
Comment (gregory-j-baker) [inline frontend/app/.server/container.init.ts:6]: Prefer aliased imports here.

File: frontend/app/.server/domain/repositories/client-application.repository.ts [hunk 2]
```diff
@@ -0,0 +1,130 @@
+import { inject, injectable } from 'inversify';
+
+import { ClientApplicationEntity } from '../entities/client-application.entity';
```
Comment (gregory-j-baker) [inline frontend/app/.server/domain/repositories/client-application.repository.ts:3]: Prefer aliased import

**Review by sebastien-comeau — DISMISSED** [review]
Nice implementation! I saw you comment about the dto. We can refactor the DTO when you will do the implementation of the service. Only return the fields that will be used and simplify the structure with good props naming convention (camelcase).

File: frontend/app/.server/constants/service-identifier.contant.ts [hunk 3]
```diff
@@ -1,5 +1,8 @@
 export const SERVICE_IDENTIFIER = {
   CLIENT_CONFIG: Symbol.for('ClientConfig'),
+  CLIENT_APPLICATION_DTO_MAPPER: Symbol.for('ClinetApplicationDtoMapper'),
+  CLIENT_APPLICATION_REPOSITORY: Symbol.for('ClinetApplicationDtoRepository'),
+  CLIENT_APPLICATION_SERVICE: Symbol.for('ClinetApplicationService'),
```
[the remark this answers is not in the document]
Comment (gregory-j-baker) [reply frontend/app/.server/constants/service-identifier.contant.ts:5]: Nice catch. lol

File: frontend/app/utils/env-utils.server.ts [hunk 4]
```diff
@@ -166,6 +166,7 @@ const serverEnv = clientEnvSchema.extend({
   LOOKUP_SVC_PREFERRED_LANGUAGE_CACHE_TTL_SECONDS: z.coerce.number().default(24 * 60 * 60),
   LOOKUP_SVC_PROVINCE_TERRITORY_STATE_CACHE_TTL_SECONDS: z.coerce.number().default(24 * 60 * 60),
   LOOKUP_SVC_PROVINCIAL_GOVERNMENT_INSURANCE_PLAN_CACHE_TTL_SECONDS: z.coerce.number().default(24 * 60 * 60),
+  CLIENT_APPLICATION_SVC_CACHE_TTL_SECONDS: z.coerce.number().default(24 * 60 * 60),
```
[the remark this answers is not in the document]
Comment (Fbasham) [reply frontend/app/utils/env-utils.server.ts:169]: i briefly thought about while doing the implementation.  My intuition says we shouldn't cache at all.  I updated the code to reflect this.

[the remark this answers is not in the document]
Comment (nicholas-ly) [reply frontend/app/utils/env-utils.server.ts]: Sounds good. We can remove `CLIENT_APPLICATION_SVC_CACHE_TTL_SECONDS` as a result right?

[the remark this answers is not in the document]
Comment (nicholas-ly) [reply frontend/app/utils/env-utils.server.ts]: Nevermind

[the remark this answers is not in the document]
Comment (Fbasham) [reply frontend/app/utils/env-utils.server.ts]: yep, just did

**Review by nicholas-ly — APPROVED** [review]

File: frontend/app/.server/domain/repositories/client-application.repository.ts [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+      throw new Error(`Failed to 'POST' for client application data. Status: ${response.status}, Status Text: ${response.statusText}`);
+    }
+
+    const data = await response.json();
+    this.log.trace('Client application: [%j]', data);
+
+    return data;
+  }
+
+  async findByFirstNameLastNameDobClientNumber(firstName: string, lastName: string, dateOfBirth: string, clientNumber: string): Promise<ClientApplicationEntity | null> {
```
Comment (sebastien-comeau) [inline frontend/app/.server/domain/repositories/client-application.repository.ts:80]: Define an object instead multiple string parameters.

Comment (sebastien-comeau) [reply frontend/app/.server/domain/repositories/client-application.repository.ts:80]:
```suggestion
  async findByFirstNameLastNameDobClientNumber(args: {firstName: string, lastName: string, dateOfBirth: string, clientNumber: string}): Promise<ClientApplicationEntity | null> {
```

File: frontend/app/.server/domain/services/client-application.service.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+import { SERVICE_IDENTIFIER } from '~/.server/constants/service-identifier.contant';
+import type { ClientApplicationDto } from '~/.server/domain/dtos/client-application.dto';
+import type { ClientApplicationDtoMapper } from '~/.server/domain/mappers/client-application.dto.mapper';
+import type { ClientApplicationRepository } from '~/.server/domain/repositories/client-application.repository';
+import type { LogFactory, Logger } from '~/.server/factories/log.factory';
+
+export interface ClientApplicationService {
+  findBySin(sin: string): Promise<ClientApplicationDto | null>;
+  findByFirstNameLastNameDobClientNumber(args: { firstName: string; lastName: string; dateOfBirth: string; clientNumber: string }): Promise<ClientApplicationDto | null>;
+}
```
Comment (sebastien-comeau) [inline frontend/app/.server/domain/services/client-application.service.ts:12]: Added `FindByFirstNameLastNameDobClientNumberSearchCriteria` definition and documentation.

```suggestion
export interface FindByFirstNameLastNameDobClientNumberSearchCriteria {
  /** The first name of the client. */
  firstName: string;
  /** The last name of the client. */
  lastName: string;
  /** The date of birth of the client in YYYY-MM-DD format. */
  dateOfBirth: string;
  /** The client number assigned to the client. */
  clientNumber: string;
}

export interface ClientApplicationRepository {
  /**
   * Finds client application data by Social Insurance Number (SIN).
   *
   * @param sin The Social Insurance Number of the client.
   * @returns A Promise that resolves to the client application data if found, or `null` if not found.
   */
  findBySin(sin: string): Promise<ClientApplicationEntity | null>;

  /**
   * Finds client application data by first name, last name, date of birth, and client number.
   *
   * @param searchCriteria An object containing the search criteria.
   * @returns A Promise that resolves to the client application data if found, or `null` if not found.
   */
  findByFirstNameLastNameDobClientNumber(searchCriteria: FindByFirstNameLastNameDobClientNumberSearchCriteria): Promise<ClientApplicationEntity | null>;
}
```

**Review by sebastien-comeau — CHANGES_REQUESTED** [review]

File: frontend/app/.server/domain/services/client-application.service.ts (hunk 6 above)
Comment (Fbasham) [reply frontend/app/.server/domain/services/client-application.service.ts]: not sure I fully understand your suggestions, but updated nonetheless.

**Review by sebastien-comeau — APPROVED** [review]

**Pull request merged**